sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,133 @@
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible ([#60067](https://github.com/kubernetes/kubernetes/pull/60067) was an attempt to do this, but it didn't seem to help too much).
+- Moving from 'endpoints' -> 'endpoint' API change (discussed in detail in below issues) would mitigate this problem as the serialization load would decrease with smaller objects.
+
+
+## Endpoints traffic is quadratic in the number of endpoints
+
+#### Issue
+
+Endpoints object for a service contains all the individual endpoints of that service. As a result, whenever even a single pod in a service is added/updated/deleted, the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers. Not being able to efficiently read/update individual endpoint changes can lead to (for e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If you consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+Overall, this is a serious performance drawback affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). The current endpoints API (which was designed at a very early stage of kubernetes when people weren't really thinking too much about scalability/performance) makes it hard to solve this problem without introducing breaking changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API with (at least) the following properties:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Separate API object for each individual endpoint
+  - Has a 'Ready' property as part of the object
+  - Port ranges and other features we’ve put off for a long time
+
+
+## Define/measure performance SLIs for endpoints propagation
+
+#### Issue
+
+We need to measure the e2e endpoints propagation latency, i.e the time from when a 
+service is created until its endpoints are populated, given that all of the associated pods 
+are already running and ready. We also need to come up with reasonable SLOs for the 
```

Comment:
I’d like to do this for pod create -> endpoint distribution latency as that is a gating factor in unidle and serverless style use of Kube

Reply:
I'm not a fan for pod created -> endpoint distribution. That contains e.g. "pod becoming ready" phase, which is completely application specific and not k8s dependent.
We already have this pod created -> pod running (not necessary ready) SLO (for ~3 years already: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos).
What is missing is the bit of pod ready -> endpoint distributed (which I would like to introduce).

Then what you're interested in is basically: (1) + (2) + some application specific time.

Anyway - I'm not opposed of measuring what you suggested too, but that isn't something you will be able to provide SLO for due to application-specific bits being part of it.

Reply:
The application specific time is not what I would measure (or keep it small to emulate a service that starts in 100ms), but since Kube's job is to run services that show up, I absolutely think 1-2 is the actual practical SLO for most users.

Reply:
So let me clarify a bit what I was trying to say.
For me, the official SLI is something that we can measure in running clusters (and is meaningful there). That pod create -> endpoint distribution latency is not really such thing. That's why I'm reluctant of making this an official SLI.

That said, I'm totally fine with measuring that in our tests and detecting regressions on that. And that's definitely a good point. I just don't think we can make an official SLI/SLO on top of that.

Reply:
So this will be addition to our tests and/or test framework, not really to the SLI.